all: add cockroachdb support

lopezator opened the pull request:
add cockroachdb support

Signed-off-by: David López <not4rent@gmail.com>

## Related PR

https://github.com/ory/hydra/pull/1326

lopezator commented:
@aeneasr I cannot figure out where is the problem here with the tests, can you give me a hand with this?

lopezator commented:
Related commit in x: https://github.com/lopezator/x/commit/44fa6032e6cd5046e0c896abb835a3bdac6b2e6f

aeneasr commented:
Thank you for all your hard work! The most recent error seems to be stemming from broken SSL?

```
--- FAIL: TestXXMigrations (302.66s)
    helper.go:58: Could not connect to database: Unable to connect to postgres (cockroach://root@localhost:26257/defaultdb?sslmode=disable): pq: SSL is not enabled on the server
```

lopezator commented:
I've doing some advances on this and been able to pass almost all tests, but still some failing. I've been working on my work to avoid spamming you. It's a draft implementation though:

If could take a look, it would be great, to collect some feedback from you.

https://github.com/lopezator/hydra/commit/d7ce922c98d56b6b68ae52c422b609a3fcd1bfd9

I'll keep you posted if I make some significative advance.

aeneasr commented:
I think it's only a couple of copy&paste errors in the circle config. Sometimes you're using `postgres` instead of `cockroach` which probably causes weirdness when connecting. Left some comments on your commit.

aeneasr commented:
So this is basically ready for a final review and then merge (after resolving conflicts)?

lopezator commented:
Not yet. I wanted to clean-up some code (ifs) I added for the cockroachdb/postgres and `sqlx` replacement.

Anyway, correct me If I am wrong but maybe I should first send a PR to `ory/x`, and then you review/merge it, as it doesn't break anything (just adds cockroach compat to x), and to remove this and point to a proper version of `ory/x` here:

https://github.com/ory/hydra/pull/1348/files#diff-37aff102a57d3d7b797f152915a6dc16R88

I'm working now on passing tests and cleaning up this on `x` here:

https://github.com/lopezator/x/commit/1b7f14d1548200310bc7fcd76feb1d1ca489e9b8

I'll send you a PR there when I'm done there.

I also opened this issue in `sqlx` and see what happens, because it would help to have a cleaner code removing the need of wrapping some `sqlx` functionality:

https://github.com/jmoiron/sqlx/issues/511

lopezator commented:
PTAL @aeneasr

lopezator commented:
Changes from latest commit:

1.- Wrapper removed, to make use of latest code in https://github.com/ory/x/pull/40
2.- Added a `driver string` param to `CreateSchema()` func as you suggested.
3.- Get the driver name from DSN in the migrate cmd.

PTAL @aeneasr

lopezator commented:
@aeneasr rebased (again) :smile:  now all green, PTAL when you could.

aeneasr commented:
Wohoo! Awesome! Checking it out now!

aeneasr commented on .circleci/config.yml:
```diff
@@ -210,6 +213,57 @@ jobs:
       - run: ./test/e2e/circle-ci.bash postgres
       - run: ./test/e2e/circle-ci.bash postgres-jwt
 
+  test-e2e-cockroach:
```
nice!

lopezator replied:
:)

aeneasr commented on Makefile:
```diff
@@ -38,14 +43,17 @@ docker:
 		./test/e2e/circle-ci.bash mysql
 		./test/e2e/circle-ci.bash mysql-jwt
+		./test/e2e/circle-ci.bash cockroach
+        ./test/e2e/circle-ci.bash cockroach-jwt
```
Indentation must be two tabs `		`

aeneasr commented on client/migrations/sql/cockroach/1.sql:
```diff
@@ -0,0 +1,39 @@
+-- +migrate Up
```
How can we ensure that each of those migration files is up to date with the latest table from the other migrations?

lopezator replied:
Aren't the tests for that? 😉. I migrated to the latest both a postgres and a cockroach database and verified the columns one by one, I'll do another pass now we stabilized the change a little bit.

aeneasr commented on client/x_manager_sql_migrations_test.go:
```diff
@@ -59,7 +60,7 @@ func TestXXMigrations(t *testing.T) {
 		migratest.MigrationSchemas{createMigrations},
 		CleanTestDB, CleanTestDB,
-		func(t *testing.T, db *sqlx.DB, _, step, steps int) {
+		func(t *testing.T, driver string, db *sqlx.DB, _, step, steps int) {
```
Maybe rename this to `dbName`, otherwise this would be considered the driver which would be expected to be `postgres` for CockroachDB.

aeneasr commented on consent/manager_sql.go:
```diff
@@ -57,9 +57,9 @@ func (m *SQLManager) PlanMigration() ([]*migrate.PlannedMigration, error) {
 }
 
-func (m *SQLManager) CreateSchemas() (int, error) {
+func (m *SQLManager) CreateSchemas(driver string) (int, error) {
```
Maybe rename this to `dbName`, otherwise this would be considered the driver which would be expected to be `postgres` for CockroachDB.

aeneasr commented on client/manager_sql.go:
```diff
@@ -231,9 +232,9 @@ func (m *SQLManager) PlanMigration() ([]*migrate.PlannedMigration, error) {
 }
 
-func (m *SQLManager) CreateSchemas() (int, error) {
+func (m *SQLManager) CreateSchemas(driver string) (int, error) {
```
Maybe rename this to `dbName`, otherwise this would be considered the driver which would be expected to be `postgres` for CockroachDB.

aeneasr commented on driver/registry_sql.go:
```diff
@@ -38,7 +36,7 @@ func init() {
 type schemaCreator interface {
-	CreateSchemas() (int, error)
+	CreateSchemas(driver string) (int, error)
 	PlanMigration() ([]*migrate.PlannedMigration, error)
```
This probably also needs a `dbName` parameter

aeneasr commented on driver/registry_sql.go:
```diff
@@ -38,7 +36,7 @@ func init() {
 
 type schemaCreator interface {
-	CreateSchemas() (int, error)
+	CreateSchemas(driver string) (int, error)
```
Maybe rename this to `dbName`, otherwise this would be considered the driver which would be expected to be `postgres` for CockroachDB.

aeneasr commented on driver/registry_sql.go:
```diff
@@ -135,32 +133,32 @@ func (m *RegistrySQL) SchemaMigrationPlan() (*tablewriter.Table, error) {
 }
 
-func (m *RegistrySQL) CreateSchemas() (int, error) {
+func (m *RegistrySQL) CreateSchemas(driver string) (int, error) {
```
Maybe rename this to `dbName`, otherwise this would be considered the driver which would be expected to be `postgres` for CockroachDB.

aeneasr commented on jwk/manager_sql.go:
```diff
@@ -71,15 +74,15 @@ func (m *SQLManager) PlanMigration() ([]*migrate.PlannedMigration, error) {
 }
 
-func (m *SQLManager) CreateSchemas() (int, error) {
+func (m *SQLManager) CreateSchemas(driver string) (int, error) {
```
Maybe rename this to `dbName`, otherwise this would be considered the driver which would be expected to be `postgres` for CockroachDB.

aeneasr commented on jwk/x_manager_sql_migrations_test.go:
```diff
@@ -58,7 +57,7 @@ func TestXXMigrations(t *testing.T) {
 		x.CleanSQL,
 		x.CleanSQL,
-		func(t *testing.T, db *sqlx.DB, k, m, steps int) {
+		func(t *testing.T, driver string, db *sqlx.DB, k, m, steps int) {
```
Maybe rename this to `dbName`, otherwise this would be considered the driver which would be expected to be `postgres` for CockroachDB.

aeneasr commented on oauth2/fosite_store_sql.go:
```diff
@@ -298,9 +301,9 @@ func (s *FositeSQLStore) deleteSession(ctx context.Context, signature string, ta
 }
 
-func (s *FositeSQLStore) CreateSchemas() (int, error) {
+func (s *FositeSQLStore) CreateSchemas(driver string) (int, error) {
```
Maybe rename this to `dbName`, otherwise this would be considered the driver which would be expected to be `postgres` for CockroachDB.

aeneasr commented on oauth2/x_fosite_migrations_test.go:
```diff
@@ -38,7 +37,7 @@ func TestXXMigrations(t *testing.T) {
 		x.CleanSQL,
 		x.CleanSQL,
-		func(t *testing.T, db *sqlx.DB, m, k, steps int) {
+		func(t *testing.T, driver string, db *sqlx.DB, m, k, steps int) {
```
Maybe rename this to `dbName`, otherwise this would be considered the driver which would be expected to be `postgres` for CockroachDB.

aeneasr commented on quickstart-cockroach.yml:
```diff
@@ -0,0 +1,36 @@
+    environment:
+      - DSN=cockroach://root@cockroachd:26257/defaultdb?sslmode=disable&max_conns=20&max_idle_conns=4
+    command:
+      migrate sql -e
```
Add `--yes` flag here which is required since the latest master.

lopezator replied:
thanks!

aeneasr commented on test/e2e/docker-compose.cockroach.yml:
```diff
@@ -0,0 +1,23 @@
+    environment:
+      - DSN=cockroach://root@cockroachd:26257/defaultdb?sslmode=disable&max_conns=20&max_idle_conns=4
+    command:
+      migrate sql -e
```
Add `--yes`

aeneasr commented on go.mod:
```diff
@@ -79,4 +80,7 @@ require (
-replace github.com/golang/lint => github.com/golang/lint v0.0.0-20190227174305-8f45f776aaf1
+replace (
+	github.com/golang/lint => github.com/golang/lint v0.0.0-20190227174305-8f45f776aaf1
+	github.com/ory/x => github.com/lopezator/x v0.0.38-0.20190423165259-d638db9fb557
```
TODO: Remove this after ory/x has been merged!

lopezator commented:
Regarding to the docs: 

What are your plans for the cockroachdb support? Release it as full and stable support from beginning or add a small note telling the user that the cockroachdb support is in a beta stage at the moment?

Anyway, we'll be testing it as soon as merged, and maybe others users could do also.

aeneasr commented:
> What are your plans for the cockroachdb support? Release it as full and stable support from beginning or add a small note telling the user that the cockroachdb support is in a beta stage at the moment?

Definitely beta/preview!

lopezator commented:
Changes applied, PTAL and tell me if you find something more to improve @aeneasr !

aeneasr commented:
Nice! So ther has been another change in the consent SQL which set a table to NOT NULL - this needs to be adopted. Then I've merged the ory/x PR so you can update it here!

aeneasr commented:
ory/x with cdb is v0.0.57

lopezator commented:
Rebased again and pointed to merged `ory/x` v0.0.57. 

Ready for final review.

lopezator commented:
PTAL when you could @aeneasr 😄  I've just rebased from master and test is broken now. Is master broken? Anything I can do?

aeneasr commented:
Hey, yeah, master is broken (my fault) - I'm currently in the middle of something but will push a fix soon. I also have this PR on my list but as you probably know, it's rather large (lot to review) so I want to give it ample time for review!

lopezator commented:
Ok, message received, thanks! 👌

aeneasr commented:
#1423

aeneasr commented on jwk/migrations/sql/tests/cockroach/1_test.sql:
```diff
@@ -0,0 +1,4 @@
+-- +migrate Up
```
Same here:

```
INSERT INTO hydra_jwk (sid, kid, version, keydata, created_at) VALUES ('4-sid', '4-kid', 0, 'some-key', NOW());
```

lopezator replied:
This has the latest contents except for the `4-` to `1-`.

aeneasr commented on oauth2/migrations/sql/tests/cockroach/1_test.sql:
```diff
@@ -0,0 +1,53 @@
+-- +migrate Up
```
Same here :)

lopezator replied:
Same, it has the contents of `9_test.sql` replacing `9-` by `1-` as it's the first migration for cockroach.

lopezator commented on client/migrations/sql/tests/cockroach/1_test.sql:
```diff
@@ -0,0 +1,4 @@
+-- +migrate Up
```
This is updated to latest `13_test.sql`, the only changes I made here are:

Replaced the string `13-` for `1-` and the order of the columns, because I have the `updated_at` and `created_at` columns as the latest.

aeneasr replied:
Ah, my bad! I think it would make sense to start migrations with `13_test.sql` and with `13-` so that we can just copy&paste those contents from the other files in the future and to keep consistency across them. What do you think? It would probably also make sense to have `13.sql` instead of `1.sql` in the main migration file, that way it's more coherent across migrations!

lopezator commented on consent/migrations/sql/tests/cockroach/1_test.sql:
```diff
@@ -0,0 +1,46 @@
+-- +migrate Up
```
I copied the contents of `consent/.../12_test.sql` and replaced the `12-` `1-`. Did I missed something here?

lopezator commented on consent/migrations/sql/cockroach/1.sql:
```diff
@@ -0,0 +1,117 @@
+-- +migrate Up
```
First manual checking should suffice for now (beta) IMHO. But if we want to ensure that indexes are 100% the same on all engines maybe we should add tests for this.

e.g. Having a list of know indexes, compare them against a list given by each DB engine.

pgsql:  `SELECT * FROM pg_indexes WHERE tablename = 'mytable';`
mysql: `SHOW INDEX FROM mytable;`
crdb:  `SHOW INDEX FROM mytable;`

lopezator commented:
PTAL @aeneasr 

I changed all the `cockroach` sql file numbers to match the `postgres` and the `mysql` migrations number, and since the tests are just inserts and now share the same index, we can safely delete the cockroach test migration sql files/folders.

I originally reordered some columns on cockroach `create table` migrations for readability, e.g. (`pk` as first column, `created_at` and `updated_at` as latest columns...). I changed that. Now the columns follow the same exact order as `mysql` and `postgres` "finished" databases, so the comparison against them is easier.

aeneasr commented:
Awesome! I think that's it - once tests pass this will be merged. Thank you for this immense effort!

aeneasr commented:
I think tests fail because we're using the same shared test sqls!

lopezator commented:
Just for the record, we freeze this until the underlying required change on `ory/x` got merged to correlate migration sql file ids, and test sql file ids.

Then, just pointing to new `ory/x` should fix the failing tests.

Just FYI, we packed `lopezator/hydra` in a container and deployed it on our cloud (to test out the cockroach support), and AFAIK it's working great.

lopezator commented:
PTAL @aeneasr 

Pending: 

1. Remove the `go.mod` replace after merging https://github.com/ory/x/pull/52
2. Squash the second commit: https://github.com/ory/hydra/pull/1348/commits/50ea15643711f07886839fbb9b179a0ae2d01b31 into the first one, I've left it like this for easier review as it's a big PR to focus only on what changed.

lopezator commented:
`go.mod` updated, rebase&squash done. Almost there! 😅

aeneasr commented:
<img width="690" alt="Screenshot 2019-05-10 at 13 17 32" src="https://user-images.githubusercontent.com/3372410/57523707-f8ffab00-7325-11e9-8b86-dd485ed682ec.png">

😎

aeneasr commented:
Thank you!

lopezator commented:
Thank you for merging and supporting me during the whole proccess! 🎉 🎉
